File: scripts/test-cst.js
const fs = require('fs');
const cstPipeline = require('../output/Nova.Compiler.CstPipeline/index.js');

// Test parsing a simple file first
const testSource = `
module Test where

import Prelude

add :: Int -> Int -> Int
add x y = x + y

data Maybe a = Nothing | Just a
`;

console.log('Testing CST pipeline...');
console.log('='.repeat(50));

const result = cstPipeline.parseModuleCst(testSource);

if (result.constructor && result.constructor.name === 'Left') {
  console.log('Parse error:', result.value0);
} else {
  console.log('Parse succeeded!');
  console.log('Module name:', result.value0.name);
  console.log('Declarations:', result.value0.declarations.length);
  for (const decl of result.value0.declarations) {
    if (decl.constructor) {
      console.log('  -', decl.constructor.name);
    }
  }
}

console.log('');
console.log('='.repeat(50));
console.log('Testing with real compiler file...');

// Try parsing the Ast.purs file
try {
  const astSource = fs.readFileSync('./src/Nova/Compiler/Ast.purs', 'utf8');
  const astResult = cstPipeline.parseModuleCst(astSource);

  if (astResult.constructor && astResult.constructor.name === 'Left') {
    console.log('Parse error on Ast.purs:', astResult.value0);
  } else {
    console.log('Ast.purs parsed successfully!');
    console.log('Module name:', astResult.value0.name);
    console.log('Declarations:', astResult.value0.declarations.length);
  }
} catch (err) {
  console.log('Error:', err.message);
}

File: nova_lang/test_cst.exs
# Test CST parsing of Types.purs
source = File.read!("../src/Nova/Compiler/Types.purs")
cst = Nova.Compiler.CstPipeline.parse_module_cst(source)

case cst do
  {:right, ast} ->
    IO.puts("Successfully parsed AST with #{length(ast.declarations)} declarations")
    
    # Show first few declarations
    ast.declarations
    |> Enum.take(20)
    |> Enum.with_index()
    |> Enum.each(fn {decl, i} ->
      name = case decl do
        {:decl_type_alias, ta} -> "TypeAlias: #{ta.name}"
        {:decl_function, f} -> "Function: #{f.name}"
        {:decl_data, d} -> "Data: #{d.name}"
        {:decl_newtype, n} -> "Newtype: #{n.name}"
        other -> "Other: #{inspect(elem(other, 0))}"
      end
      IO.puts("  #{i}: #{name}")
    end)
    
  {:left, err} ->
    IO.puts("Parse error: #{inspect(err)}")
end
